Added support for A/B testing and staged rollouts of new functionality

karrtikr opened the pull request:
For #5042 

<!--
  If an item below does not apply to you, then go ahead and check it off as "done" and strikethrough the text, e.g.:
    - [x] ~Has unit tests & system/integration tests~
-->
- [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [x] Title summarizes what is changing
- [x] Has a [news entry](https://github.com/Microsoft/vscode-python/tree/master/news) file (remember to thank yourself!)
- [x] Appropriate comments and documentation strings in the code
- [x] Has sufficient logging.
- [x] Has telemetry for enhancements.
- [x] Unit tests & system/integration tests are added/updated
- ~[ ] [Test plan](https://github.com/Microsoft/vscode-python/blob/master/.github/test_plan.md) is updated as appropriate~
- ~[ ] [`package-lock.json`](https://github.com/Microsoft/vscode-python/blob/master/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed)~
- ~[ ] The wiki is updated with any design decisions/details.~

DonJayamanne commented on src/client/common/crypto.ts:
```diff
@@ -0,0 +1,17 @@
+@injectable()
+export class CryptoUtils implements ICryptoUtils {
+
+    public async createHash(algorithm: string, data: string, encoding: HexBase64Latin1Encoding, hashFormat: 'number' | 'string'): Promise<number | string | Buffer> {
```
Please remove the `algorithm` argument, I don't see this ever changing.
Lets always assume we will use `sha256` for hasing.

DonJayamanne replied:
Why is this method `async`?

DonJayamanne commented on src/client/common/types.ts:
```diff
@@ -399,6 +399,11 @@ export interface IAsyncDisposable {
 
+export const ICryptoUtils = Symbol('ICryptoUtils');
+export interface ICryptoUtils {
+    createHash(algorithm: string, data: string, encoding: string, hashFormat: 'number' | 'string'): Promise<number | string | Buffer>;
```
Why return a `Buffer`?
Do we have places in our code where we use this? If not, then please leave as `number | string`.

There's a problem with this type.
When the caller of the method passes in a hashFormat of type 'number', the return value is not a `number`, i.e. its not strongly typed!! This is a problem.
Casting the type to a specific version works, but that's a work around and not guaranteed to work. E.g. tomorrow we change the return value to `Object` or some interface, then the compiler will not throw any errors as we're explicitly casting to a specific version.

**Solution, please use conditional types, see below:**
https://www.typescriptlang.org/docs/handbook/advanced-types.html

```typescript
export interface IHashFormat {
    'number': number;
    'string': string;
}
export interface ICryptoUtils
	createHash<P extends IHashFormat, E extends keyof P>(data: string, encoding: string, hashFormat: E): Promise<E extends 'number' ? number : string>;
}

# Implementation
export class CryptoUtils implements ICryptoUtils {
	public createHash<P extends IHashFormat, E extends keyof P>(data: string, encoding: string, hashFormat: E): Promise<E extends 'number' ? number : string>;
		return hashFormat === 'number' ? +hash : hash as any;
	}
}
```

karrtikr replied:
Had to use `E extends keyof IHashFormat` instead of `P extends IHashFormat, E extends keyof P`, otherwise the following error popped up (not sure why):
`Types of hashFormat in interface and class are not compatible`

Also, simplified the return type.

DonJayamanne commented on src/client/activation/experiments.ts:
```diff
@@ -0,0 +1,76 @@
+
+const EXPIRY_DURATION_MS = 30 * 60 * 1000;
+const experimentTimestampKey = 'EXPERIMENT_TIMESTAMP_KEY';
+const experimentsLocation = path.join(EXTENSION_ROOT_DIR_FOR_TESTS, 'src', 'client', 'experiments.json');
```
Please use the persistence storage service to store the json.
We should avoid creating files.

DonJayamanne commented on src/client/activation/experiments.ts:
```diff
@@ -0,0 +1,76 @@
+    @traceDecorators.error('Failed to download experiments file')
+    protected async downloadExperiments() {
+        this.experiments = await this.httpClient.getJSON('https://raw.githubusercontent.com/karrtikr/check/master/environments.json');
+        await this.fs.writeFile(experimentsLocation, JSON.stringify(this.experiments, null, 4));
```
Please save JSON into persistence storage instead of creating a file.
Use the solution we already have today.

karrtikr replied:
I've implemented that, although for some reason storing JSON in persistent state is having flaky behavior.
Sometimes, it does not get stored which leads to re-downloading the file, and i am not able to figure out why.

DonJayamanne commented on src/client/activation/experiments.ts:
```diff
@@ -0,0 +1,76 @@
+    @traceDecorators.error('Failed to download experiments file')
+    protected async downloadExperiments() {
+        this.experiments = await this.httpClient.getJSON('https://raw.githubusercontent.com/karrtikr/check/master/environments.json');
+        await this.fs.writeFile(experimentsLocation, JSON.stringify(this.experiments, null, 4));
```
* Please use `downloadFile` that'll download to a temporary location
* Parse file using `jsonc-parser` instead of using `JSON`
* Store the JSON

We need to download the file and parse manually as the contents of the file could contain comments and trailing commas (meaning it could be `JSONC`).

karrtikr replied:
Using `downloadFile` led to a code looking like this, which seems a [tad tedious](https://nodejs.org/en/docs/guides/anatomy-of-an-http-transaction/#request-body).

```
protected async downloadExperiments() {
    const req = await this.httpClient.downloadFile(configUri);
    const body: Buffer[] = [];
    let text: string;
    req
        .on('response', (response) => {
            if (response.statusCode !== 200) {
                traceError(`Failed with status ${response.statusCode}, ${response.statusMessage}, Uri ${configUri}`);
            }
        })
        .on('error', (error) => {
            traceError('Request failed', error);
        })
        .on('data', (chunk: any) => {
            body.push(chunk);
        }).on('end', () => {
            text = Buffer.concat(body).toString();
            this.experiments = parse(text, [], { allowTrailingComma: true, disallowComments: false });
        });
}
```

Plus the code proceeds without downloading, i.e Function `downloadExperiments()` doesn't wait for the download to finish. 

It makes sense to just add another method to **get json without trailing commas and comments** in `httpClient.ts`

DonJayamanne commented on src/client/activation/experiments.ts:
```diff
@@ -0,0 +1,76 @@
+        if (index < 0) {
+            return false;
+        }
+        const hash = await this.crypto.createHash('sha512', `${vscode.env.machineId}+${this.experiments[index].salt}`, 'hex', 'number') as number;
```
Please use `IApplicationEnvironment.machineId` instead of `vscode.env.machineId`

DonJayamanne commented on src/client/activation/experiments.ts:
```diff
@@ -0,0 +1,76 @@
+        this.experiments.forEach(experiment => sendTelemetryEvent(EventName.PYTHON_EXPERIMENTS, undefined, experiment));
+    }
+
+    public async inExperiment(experimentName: string): Promise<boolean> {
```
Add a try catch around the code in this function and return a default value of `false` if there are any exceptions.
Remember we don't want extension to fall over just because something is wrong in the extension json or other.

DonJayamanne replied:
This is part of the requirement, we need this code to be as resilient as possible.

karrtikr replied:
One of the reasons why this is a WIP :P Thanks for reminding anyways.

DonJayamanne commented on src/client/activation/activationManager.ts:
```diff
@@ -58,6 +59,11 @@ export class ExtensionActivationManager implements IExtensionActivationManager {
     }
     protected async initialize() {
+        try {
+            await this.experiments.initialize();
```
Please remove this code from here and ensure ExperimentManager implements `IExtensionActivationService`. Once registered, it'll get auto instantiated.

DonJayamanne commented on src/client/activation/experiments.ts:
```diff
@@ -0,0 +1,76 @@
+        @inject(ICryptoUtils) private readonly crypto: ICryptoUtils
+    ) { }
+
+    public async initialize() {
```
We don't want this `initialize` method to block anything. 
Assume user has a very slow network connection, we don't want that to block loading of the extension.

**Solution:**
* Initialize in the background

```typescript
public async ininitialize(){
	// This method will return immediately, hence not blocking the extension.
	initializeBackground().ignoreErrors();
}

@traceDecorators.error('Failed to initialize experiments')
public async initializeBackground(){
	// Usual code...
}
```

karrtikr replied:
Agree, although shouldn't we have an experiments file before launching the language server or jedi? I did it this way to ensure this, but i understand your point.

DonJayamanne replied:
Experiments will not be shipped with extension.

DonJayamanne commented on src/client/activation/experiments.ts:
```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Please move this file from `activation` folder into `common`

DonJayamanne commented on src/client/activation/types.ts:
```diff
@@ -126,3 +126,9 @@ export interface IPlatformData {
+export const IExperimentsManager = Symbol('IExperimentsManager');
+export interface IExperimentsManager {
+    initialize(): Promise<void>;
+    inExperiment(experimentName: string): Promise<boolean>;
```
Please make this a synchronous function.
We want `inExperiment` to be fast.
It could be used in a number of places and we don't want this slowing this down.

DonJayamanne commented on src/client/activation/experiments.ts:
```diff
@@ -0,0 +1,76 @@
+            return;
+        }
+        await IsFileValid.updateValue(true);
+        this.experiments.forEach(experiment => sendTelemetryEvent(EventName.PYTHON_EXPERIMENTS, undefined, experiment));
```
Do we need this?

karrtikr replied:
I did in reference to this line in the mail, `Log telemetry event with all of the experiments enabled for that session`.

What type of telemetry should i send corresponding to this line?

DonJayamanne replied:
That needs to be done when the extension starts for the control group, and when a user is identified as belonging to another group.
Lets discuss later

DonJayamanne commented on src/client/activation/experiments.ts:
```diff
@@ -0,0 +1,76 @@
+        if (this.isTelemetryDisabled()) {
+            return;
+        }
+        const IsFileValid = this.persistentStateFactory.createGlobalPersistentState(experimentTimestampKey, false, EXPIRY_DURATION_MS);
```
lower case `I`.
Should be `const isFileValid`

DonJayamanne commented on src/client/activation/experiments.ts:
```diff
@@ -0,0 +1,76 @@
+
+    @traceDecorators.error('Failed to download experiments file')
+    protected async downloadExperiments() {
+        this.experiments = await this.httpClient.getJSON('https://raw.githubusercontent.com/karrtikr/check/master/environments.json');
```
Please use a constant.

karrtikr commented on src/client/activation/experiments.ts:
```diff
@@ -0,0 +1,76 @@
+        try {
+            await this.downloadExperiments();
+        } catch (err) {
+            sendTelemetryEvent(EventName.PYTHON_EXPERIMENTS, undefined, { error: 'Failed to download experiments file' }, err);
```
I need to continue execution to the next line (`await IsFileValid.updateValue(true)`) only if the download succeeds.
From what i understand `@swallowExceptions` will continue the execution regardless of success or failure.

DonJayamanne replied:
Yes agreed.

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,96 @@
+            }
+            const hash = this.crypto.createHash(`${this.appEnvironment.machineId}+${this.experiments[index].salt}`, 'hex', 'number');
+            const inExp = hash % 100 >= this.experiments[index].min && hash % 100 < this.experiments[index].max;
+            if (inExp) {
```
No point logging this here.
The requirement is to log this as soon as extension loads.
This is too late.

DonJayamanne replied:
The code that logs the list of experiment names the user belongs must go into the above `activate` method.
This will ensure we log this early on when extension starts.

DonJayamanne replied:
Note, logging of telemetry MUST happen only when we use the method `inExperiment`, i.e. the time when external code attempts to use this API.

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,96 @@
+            return;
+        }
+        try {
+            await this.downloadExperiments();
```
Do we need this `try...catch`.
If this method fails, then the whole `initializeInBackground` method fails, so do we need a `return`. No we don't.

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,96 @@
+
+    public inExperiment(experimentName: string): boolean {
+        try {
+            const experimentStorage = this.persistentStateFactory.createGlobalPersistentState(experimentStorageKey, undefined as any, EXPIRY_DURATION_MS);
```
Why are we doing this again here, this same code is in `initializeInBackground`.

DonJayamanne replied:
Instead of using two state stores, the `experimentStorage` and the property please use one.
Right now you have code to keep both in sync this can be avoided if you removed `this.experiments` completely.

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,96 @@
+            }
+            const experimentNames = this.experiments.map(experiment => experiment.name);
+            const index = experimentNames.indexOf(experimentName);
+            if (index < 0) {
```
You can collapse all of the above into a simple `this.experiments.find`

DonJayamanne commented on src/client/common/net/httpClient.ts:
```diff
@@ -37,4 +38,19 @@ export class HttpClient implements IHttpClient {
         });
     }
+    public async getJSONC<T>(uri: string, options?: ParseOptions | undefined): Promise<T> {
+        // tslint:disable-next-line:no-require-imports
```
Why do we need the `options` argument?
Please remove it and use the default of `allow comments and allow trailing commas`.
If we want to change this, then we can add an argument

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,96 @@
+    ) { }
+
+    public async activate(resource: Uri): Promise<void> {
+        this.resource = resource;
```
We need to activate this class only once per workspace.
Please ensure we have a guard against multi-root workspace activations.

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,91 @@
+const experimentStorageKey = 'EXPERIMENT_STORAGE_KEY';
+const configUri = 'https://raw.githubusercontent.com/karrtikr/check/master/environments.json';
+
+type AB_Experiments = { name: string; salt: string; min: number; max: number }[];
```
Please remove the `_`, TypeScript doesn't use `_` in type names.
```suggestion
type ABExperiments = { name: string; salt: string; min: number; max: number }[];
```

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,91 @@
+        if (this.isTelemetryDisabled() || this.experimentStorage.value) {
+            return;
+        }
+        const downloadedExperiments = await this.httpClient.getJSONC<AB_Experiments>(configUri);
```
Remove `_` from typenames (TypeScript doesn't use `_` in typenames)
```suggestion
        const downloadedExperiments = await this.httpClient.getJSONC<ABExperiments>(configUri);
```

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,91 @@
+type AB_Experiments = { name: string; salt: string; min: number; max: number }[];
+@injectable()
+export class ExperimentsManager implements IExperimentsManager {
+    private experimentStorage: IPersistentState<AB_Experiments | undefined>;
```
```suggestion
    private experimentStorage: IPersistentState<ABExperiments | undefined>;
```

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,91 @@
+        @inject(IApplicationEnvironment) private readonly appEnvironment: IApplicationEnvironment,
+        @inject(IOutputChannel) @named(STANDARD_OUTPUT_CHANNEL) private readonly output: IOutputChannel
+    ) {
+        this.experimentStorage = this.persistentStateFactory.createGlobalPersistentState<AB_Experiments | undefined>(experimentStorageKey, undefined, EXPIRY_DURATION_MS);
```
```suggestion
        this.experimentStorage = this.persistentStateFactory.createGlobalPersistentState<ABExperiments | undefined>(experimentStorageKey, undefined, EXPIRY_DURATION_MS);
```

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,91 @@
+                return true;
+            }
+        } catch (ex) {
+            traceError('Failed to check if user is in experiment', ex);
```
```suggestion
            traceError(`Failed to check if user is in experiment '{experimentName}'`, ex);
```

karrtikr commented:
@DonJayamanne  @brettcannon 
Problems I notice right now,
1. For a user in say "LS - enabled" experiment, consider the following setup:

**_Sets clock to zero_**
- _t = 0_ : Jedi Python language engine is activated the first time, because experiments are still downloading.

- _t = 25_ : If the user activates the extension again within the next 30 minutes (say 25th minute), only then LS will be activated for them.

- _t = 40_ : User activates the extension after 30 minutes (say 40th minute), again Jedi will be activated as experiments will still need to be downloaded.

So even for a "LS - enabled" user, he will keep switching back and forth between LS and jedi, which is not a good experience (not good for the experiment either).

**SOLUTION:**  Should increase time gap between downloads.

2. **Implementation problem :** Storing experiments in persistent storage seems to be flaky,
`await this.experimentStorage.updateValue(downloadedExperiments);`
Sometimes updates storage, and other times doesn't. Storing experiments in file worked everytime.

ericsnowcurrently commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,109 @@
+import { swallowExceptions } from './utils/decorators';
+import { Experiments } from './utils/localize';
+
+const EXPIRY_DURATION_MS = 30 * 60 * 1000;
```
So we're updating the "downloaded" experiments every 30 minutes?

karrtikr replied:
Yes, that's what the requirement specified.

ericsnowcurrently commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    public inExperiment(experimentName: string): boolean {
+        try {
```
Shouldn't we block here until `initializeInBackground()` has finished?

karrtikr replied:
We don't want to slow activating the extension if user has a low speed internet. If the experiments config file is not present, it will get downloaded and used up in the next startup.

ericsnowcurrently commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,109 @@
+
+    public inExperiment(experimentName: string): boolean {
+        try {
+            const experiments = this.experimentStorage.value ? this.experimentStorage.value : [];
```
I recommend a comment here pointing out that `experimentStorage` is populated in `ExtensionActivationManager.activateWorkspace()`.

ericsnowcurrently commented on src/client/common/types.ts:
```diff
@@ -399,7 +401,25 @@ export interface IAsyncDisposable {
     push(disposable: IDisposable | IAsyncDisposable): void;
 }
+
+export type ABExperiments = { name: string; salt: string; min: number; max: number }[];
```
It would be *super* helpful to have a comment here explaining what the `ABExperiments` field mean.  Bonus points for a link to a wiki page on the design of the A/B functionality.

ericsnowcurrently commented on src/client/common/types.ts:
```diff
@@ -399,7 +401,25 @@ export interface IAsyncDisposable {
     push(disposable: IDisposable | IAsyncDisposable): void;
 }
+
+export type ABExperiments = { name: string; salt: string; min: number; max: number }[];
```
Also, shouldn't the name be singular (`ABExperiment`)?

karrtikr replied:
It's an array of experiments.

ericsnowcurrently replied:
Ah, I missed the array part. :)

ericsnowcurrently commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,109 @@
+        return false;
+    }
+
+    public isTelemetryDisabled(): boolean {
```
Why isn't this part of the code in `src/client/common/telemetry/index.ts`?

karrtikr replied:
I just noticed this line under the function,
`this.workspaceService.getConfiguration('telemetry', this.resource).inspect<boolean>('enableTelemetry')!;`

If I move this over to `index.ts`,

- I will have to use `vscode.workspace.getConfiguration`, which should be okay.
- But I won't be able to test this, and will have to discard the tests I have already written.

What do you recommend?

ericsnowcurrently replied:
I recommend putting it with the other telemetry code.

karrtikr replied:
After discussion with Don, the function `isTelemetryDisabled()` has become experiment-specific. For the purpose of just our experiments, we are just checking the user settings, assuming users always use that to disable telemetry.

As it will not be checking workspace folder and workspace settings, moving it to general telemetry section does not seem appropriate.

ericsnowcurrently commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
```
This would be a great place for a comment that includes a link back to a wiki page that explains about our A/B functionality.

ericsnowcurrently commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    @traceDecorators.error('Failed to initialize experiments')
+    public async initializeInBackground() {
```
How is this reliably tested (without races) if it's happening in the background?

karrtikr replied:
Good point... will add a unit test for that.

ericsnowcurrently commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,109 @@
+            (settings.workspaceFolderValue !== true && settings.workspaceValue !== true && settings.globalValue === false)) ? true : false;
+    }
+
+    private isUserInRange(min: number, max: number, salt: string) {
```
This is technically complex enough that it deserves dedicated unit tests, to be sure we've gotten the details right.

karrtikr replied:
It is being tested, as part `inExperiment()`.

ericsnowcurrently replied:
Let's discuss.

DonJayamanne replied:
> I understand. But it is being tested, as part of testing inExperiment(). Testing it separately would lead to duplication. What do you think?

I agree with @ericsnowcurrently And @karrtikr you might remember, I did suggest putting this into a separate class. (my argument was, we have similar code elsewhere).
Testing this logic as part of a method `inExperiment` isn't obvious enough and doesn't feel right.

DonJayamanne replied:
However I leave it upto you based on your discussion with @ericsnowcurrently

ericsnowcurrently commented:
@karrtikr, be sure to ask me if you have any questions about my review.

karrtikr commented on src/client/activation/activationService.ts:
```diff
@@ -113,6 +114,25 @@ export class LanguageServerExtensionActivationService implements IExtensionActiv
+    }
+
+    public useJedi(): boolean {
+        if (this.isJediUsingDefaultConfiguration()) {
```
Yes, it's just a use case.

karrtikr replied:
I actually did this to test the functionality added.

karrtikr commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,109 @@
+const isStorageValidKey = 'IS_EXPERIMENTS_STORAGE_VALID_KEY';
+export const experimentStorageKey = 'EXPERIMENT_STORAGE_KEY';
+export const downloadedExperimentStorageKey = 'DOWNLOADED_EXPERIMENTS_STORAGE_KEY';
+const configUri = 'https://raw.githubusercontent.com/karrtikr/check/master/environments.json';
```
I am thinking to commit `environments.json` to the `vscode-python` repo, and then link it here, sounds good?

ericsnowcurrently replied:
This is more of a design decision.  Potentially it should be hosted somewhere else.  So I suggest bringing it up with the team.

karrtikr commented on src/test/common/experiments.unit.test.ts:
```diff
@@ -0,0 +1,202 @@
+                expectedResult: false
+            },
+            {
+                testName: 'If experiment\'s name is in experiment list and hash modulo output is in range, user is in experiment',
```
`isUserInRange` is being tested in the following 3 tests.

karrtikr commented:
@ericsnowcurrently Thanks for the review!
> 
> * the change in `ExtensionActivationManager` is pretty minimal so why is there so much churn in the corresponding tests file?

They are mostly white space changes, you can hide whitespace changes in the change tab.
> * it seems like the test coverage isn't great

I have to disagree here.. as every line of the added code is hit atleast once.
> * it's really hard to follow what's going on in the tests (this is at least partly due to how we do testing in the extension)

Let's discuss about this, I really want to understand how you see this and your suggestions.
> * most of the new tests looks more like mid-level integration tests, rather than unit tests
> * no functional tests

The way I see it I am testing every function one by one, there is one `private` function `isUserInRange` which I am leaving out, as it is already being tested thoroughly in `inExperiment` ([here](https://github.com/microsoft/vscode-python/pull/5530/files#r290079102)). I can unit test it separately but that just leads to duplication of code.

ericsnowcurrently commented on src/client/activation/activationService.ts:
```diff
@@ -113,6 +114,25 @@ export class LanguageServerExtensionActivationService implements IExtensionActiv
+            if (this.abExperiments.inExperiment('LS - enabled')) {
+                return false;
+            }
+            this.abExperiments.inExperiment('LS - control');
```
What's the point of this line?  If telemetry then please add a comment to that effect.

karrtikr replied:
Yes, it's telemetry for the control group. Will add.

DonJayamanne replied:
Please add a constant for `LS - control`. Move these names into a `experiments.ts` or `experimentGroups.ts` or similar. We dont want these scattered all over the place. they must be defined in one place.

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,112 @@
+
+    @traceDecorators.error('Failed to initialize experiments')
+    public async initializeInBackground() {
+        if (this.isTelemetryDisabled() || this.isStorageValid.value) {
```
Why do we need another property `isStorageValid`?
Why can't we just check whether `downloadedExperimentsStorage.value` is undefined or not?

Its easier to test and read as well. Currently we're using an extra property unnecessarily.

karrtikr replied:
This is done for a very particular reason. Consider the following case if we do not use `isStorageValid`,

- t=0: User activates the extension, experiments get downloaded into `downloadedExperimentsStorage`. (`experimentsStorage` is still empty)
- t>30 minutes: User again activates the extension, previously downloaded experiments are NOT used as they have expired, so the same experiments are downloaded again.

User will never be able to use the experiments.

DonJayamanne commented on src/test/common/experiments.unit.test.ts:
```diff
@@ -0,0 +1,235 @@
+        experimentStorage = TypeMoq.Mock.ofType<IPersistentState<any>>();
+        downloadedExperimentsStorage = TypeMoq.Mock.ofType<IPersistentState<any>>();
+        output = TypeMoq.Mock.ofType<IOutputChannel>();
+        when(persistentStateFactory.createGlobalPersistentState(anything(), false, anything())).thenReturn(isStorageValid.object);
```
Why is the first argument of `createGlobalPersistentState` `anything`? Please use a specific value.

karrtikr replied:
For now, this works fine as there is only ONE call to `createGlobalPersistentState` which uses 3 arguments, `createGlobalPersistentState(anything(), false, anything()))`.

I have seen plenty of places in the extension where `anything()` is used if there is no ambiguity...

DonJayamanne commented on src/client/activation/types.ts:
```diff
@@ -42,7 +42,7 @@ export interface ILanguageServerActivator extends IDisposable {
 export interface IHttpClient {
     downloadFile(uri: string): Promise<RequestResult>;
-    getJSON<T>(uri: string): Promise<T>;
+    getJSON<T>(uri: string, strict?: boolean): Promise<T>;
```
If we want to collapse `JSON` into `JSONC`, then please add comments into the interface for the arguments.
I.e. what's the implication of using `strict`?

karrtikr replied:
Added the description on top of function definition

DonJayamanne replied:
The docs must be added to the interface, not the implementation.

DonJayamanne commented on src/client/activation/activationService.ts:
```diff
@@ -113,6 +114,25 @@ export class LanguageServerExtensionActivationService implements IExtensionActiv
+
+    public useJedi(): boolean {
+        if (this.isJediUsingDefaultConfiguration()) {
+            if (this.abExperiments.inExperiment('LS - enabled')) {
```
Please add a constant for `LS - control`. Move these names into a `experiments.ts` or `experimentGroups.ts` or similar. We dont want these scattered all over the place. they must be defined in one place.

DonJayamanne commented on src/client/common/types.ts:
```diff
@@ -399,7 +401,29 @@ export interface IAsyncDisposable {
+export type ABExperiments = { name: string; salt: string; min: number; max: number }[];
+export const IExperimentsManager = Symbol('IExperimentsManager');
+export interface IExperimentsManager extends IExtensionActivationService {
+    experimentStorage: IPersistentState<ABExperiments | undefined>;
```
Why are internal states public? These must be private

DonJayamanne commented on src/client/common/types.ts:
```diff
@@ -399,7 +401,29 @@ export interface IAsyncDisposable {
+export const IExperimentsManager = Symbol('IExperimentsManager');
+export interface IExperimentsManager extends IExtensionActivationService {
+    experimentStorage: IPersistentState<ABExperiments | undefined>;
+    downloadedExperimentsStorage: IPersistentState<ABExperiments | undefined>;
```
Why are internal states public? These must be private

DonJayamanne commented on src/client/common/types.ts:
```diff
@@ -399,7 +401,29 @@ export interface IAsyncDisposable {
+export interface IExperimentsManager extends IExtensionActivationService {
+    experimentStorage: IPersistentState<ABExperiments | undefined>;
+    downloadedExperimentsStorage: IPersistentState<ABExperiments | undefined>;
+    inExperiment(experimentName: string): boolean;
```
Instead of retuning `boolean` please return `boolean | undefined`.
When `true` we know user is in an experiement.
When `false` we know user is **definitely not** in an experiement.
When `undefined` we don't know (e.g. in the case experiments have not loaded, etc...).

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,112 @@
+        @inject(IApplicationEnvironment) private readonly appEnvironment: IApplicationEnvironment,
+        @inject(IOutputChannel) @named(STANDARD_OUTPUT_CHANNEL) private readonly output: IOutputChannel
+    ) {
+        this.isStorageValid = this.persistentStateFactory.createGlobalPersistentState<boolean>(isStorageValidKey, false, EXPIRY_DURATION_MS);
```
Also I don't think the logic is correct for the expiry duration. 
Are we supposed to check for new experiments every 30 minutes?

karrtikr replied:
I don't get the issue here. If extension has not activated within the 30 minutes, update the experiments.

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,112 @@
+        @inject(IOutputChannel) @named(STANDARD_OUTPUT_CHANNEL) private readonly output: IOutputChannel
+    ) {
+        this.isStorageValid = this.persistentStateFactory.createGlobalPersistentState<boolean>(isStorageValidKey, false, EXPIRY_DURATION_MS);
+        this.experimentStorage = this.persistentStateFactory.createGlobalPersistentState<ABExperiments | undefined>(experimentStorageKey, undefined);
```
Why do we have two properties `experimentStorage` & `downloadedExperimentsStorage`.
They both seem to store the exact same thing! Why?

DonJayamanne replied:
Why can't we just remove `downloadedExperimentsStorage` and use `experimentStorage` alone. I don't see why it wouldn't work. I.e. why the extra hop.

karrtikr replied:
Lets discuss... I did this to ensure all experiments group are logged just when the experiment starts.

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,112 @@
+        }
+        this.activatedWorkspaces.set(this.getWorkspacePathKey(resource), true);
+        this.resource = resource;
+        this.logExperimentGroups();
```
If we have multiple workspaces, then the method `logExperimentGroups` gets executed multiple times? That's wrong

karrtikr replied:
Good point! Thanks.

DonJayamanne commented on src/client/activation/activationManager.ts:
```diff
@@ -54,6 +55,11 @@ export class ExtensionActivationManager implements IExtensionActivationManager {
+        if (Array.isArray(this.abExperiments.downloadedExperimentsStorage.value)) {
+            await this.abExperiments.experimentStorage.updateValue(this.abExperiments.downloadedExperimentsStorage.value);
+            await this.abExperiments.downloadedExperimentsStorage.updateValue(undefined);
+        }
```
On second thought, I don't see the need for this code!
Why can't this be done in the `activate` method of the extension manager.

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,112 @@
+    private logExperimentGroups(): void {
+        if (Array.isArray(this.experimentStorage.value)) {
+            for (const exp of this.experimentStorage.value) {
+                if (this.inExperiment(exp.name)) {
```
Doing this will result in telemetry being sent out. That's wrong.
E.g. assume user installed the extension the first time and loaded it.
As the experiments are being downloaded, the language server will not get used & user won't be in the `LS - enabled` experiment group. This is because the experiments are still getting downloaded.

After a few seconds/minute or two, experiments get downloaded.
Now we execute the following code an send telemetry indicating the fact that the user is in the experiment `LS - enabled` and the are using `Language Server`. 

This will lead to incorrect information being sent up.

karrtikr replied:
I agree that this will be sending Telemetry as well (which it should not be doing), but it won't be sending incorrect ones.
Because this is not how it works.. downloaded experiments are NEVER used in the current session.

DonJayamanne commented on src/client/activation/activationService.ts:
```diff
@@ -113,6 +115,26 @@ export class LanguageServerExtensionActivationService implements IExtensionActiv
+                return false;
+            }
+            // Send telemetry for the control group
+            this.abExperiments.inExperiment(LSControl);
```
This is a method named `isExperiment` used to check whether user is in an experiment.
Here we're calling this method and not using the return value, but calling it knowing the internals of it, because internally the method sends telemetry!
I.e. the method isn't clear enough, you need to know the internals of the method to use this.

My suggestion, create a separate method that'll be more specific to send telemetry.

ericsnowcurrently replied:
Shouldn't we handle the `false` case?  Otherwise not just send telemetry (instead of also doing the checks)?

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,154 @@
+
+    /**
+     * Updates experiment storage using local data if available
+     */
```
Please add comments about why we are doing this. Why do we need the two state stores.
This isn't obvious here.
The comments further up also don't make this obvious.

DonJayamanne commented on src/client/common/net/httpClient.ts:
```diff
@@ -22,7 +23,13 @@ export class HttpClient implements IHttpClient {
+    /**
+     * Downloads file from uri as string and parses them into JSON objects
+     * @param uri The uri to download the JSON from
+     * @param strict Set `false` to allow trailing comma and comments in the JSON, defaults to `true`
```
Doc strings belong to interfaces. No one users the concrete classes.

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,154 @@
+        }
+
+        // Step 2. Update experiment storage using local experiments file if available
+        if (await this.fs.fileExists(configFile)) {
```
What's the purpose of this `config` file?
I didn't think we'd ship with default experiments!

karrtikr replied:
Everybody came to this consensus in the meeting with Nicholas...
This is done to ensure that experiments are used in the first session itself

ericsnowcurrently commented on src/test/activation/activationService.unit.test.ts:
```diff
@@ -97,405 +105,478 @@ suite('Activation - ActivationService', () => {
+                    experiments
+                        .setup(ex => ex.inExperiment(TypeMoq.It.isAny()))
+                        .returns(() => false)
+                        .verifiable(TypeMoq.Times.atLeastOnce());
```
If `jediIsEnabled` is *not* `undefined` then why would `inExperiment()` ever be called?  Up in `useJedi()`, `isJediUsingDefaultConfiguration()` should have returned `false`.  So `inExperiment()` should never be called.

karrtikr replied:
Forgot to remove it with all the refactoring going on, thanks.

ericsnowcurrently commented on src/test/activation/activationService.unit.test.ts:
```diff
@@ -97,405 +105,478 @@ suite('Activation - ActivationService', () => {
-                activator.verifyAll();
-                serviceContainer.verifyAll();
-            }
+            if (jediIsEnabled !== undefined) {
```
I'd expect all these tests to be exercised for `undefined` as well, which should be treated the same a `false`, right?  In that case, I *would* expect `inExperiment()` to be called once.

The tests for `jediIsEnabled === undefined`, combined with `inExperiment()` returning `true`, should be tested separately from the "Test default jedi value" suite, rather than as a case in it.

ericsnowcurrently commented on src/test/activation/activationService.unit.test.ts:
```diff
@@ -97,405 +105,478 @@ suite('Activation - ActivationService', () => {
+                                .verifiable(TypeMoq.Times.once());
+
+                            const activationService = new LanguageServerExtensionActivationService(serviceContainer.object, stateFactory.object, experiments.object);
+                            const result = activationService.isJediUsingDefaultConfiguration(Uri.parse('a'));
```
Nearly all the other tests call `activationService.activate()`, so it's easy to miss that something else is getting tested here.

ericsnowcurrently commented on src/test/activation/activationService.unit.test.ts:
```diff
@@ -97,405 +105,478 @@ suite('Activation - ActivationService', () => {
+                        }
+                    ];
+
+                suite('Function isJediUsingDefaultConfiguration()', () => {
```
With all this nesting of suites, it's kind of hard to follow what is getting tested under which conditions.  I recommend having separate top-level suites for the distinct things you are testing.

DonJayamanne replied:
Agreed

ericsnowcurrently commented on src/test/activation/activationService.unit.test.ts:
```diff
@@ -97,405 +105,478 @@ suite('Activation - ActivationService', () => {
-                    activator2.verifyAll();
+
+                    const activationService = new LanguageServerExtensionActivationService(serviceContainer.object, stateFactory.object, experiments.object);
+                    const result = activationService.useJedi();
```
Nearly all the other tests call activationService.activate(), so it's easy to miss that something else is getting tested here.

karrtikr replied:
Refactored the tests

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,154 @@
+        }
+
+        // Step 2. Update experiment storage using local experiments file if available
+        if (await this.fs.fileExists(configFile)) {
```
Personally I don't see the need for this. Not sure this is part of the requirement.

karrtikr replied:
> Finally, don't we want to leave the current experiments in place if nothing got downloaded the last time? With this code, the current experiments will be replaced with the default...

Good point. Making sure that the experiment file can only be used ONCE should ensure this does not happen.

ericsnowcurrently commented on src/test/activation/activationService.unit.test.ts:
```diff
@@ -97,405 +105,478 @@ suite('Activation - ActivationService', () => {
+                    const settings = {};
+                    experiments
+                        .setup(ex => ex.inExperiment('LS - enabled'))
+                        .returns(() => true)
```
There should be tests for the "return false" case.

DonJayamanne commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,154 @@
+        this.activatedOnce = true;
+        await this.updateExperimentStorage();
+        this.logExperimentGroups();
+        this.initializeInBackground().ignoreErrors();
```
Curious, why is that better?

ericsnowcurrently commented on src/test/activation/activationService.unit.test.ts:
```diff
@@ -97,405 +105,478 @@ suite('Activation - ActivationService', () => {
+                test('If LS-enabled experiment is enabled and default value of jedi is being used, then LS is enabled', async () => {
+                    const settings = {};
+                    experiments
+                        .setup(ex => ex.inExperiment('LS - enabled'))
```
What about tests where we ensure it gets called with "LS - control"?

ericsnowcurrently commented on src/test/telemetry/index.unit.test.ts:
```diff
@@ -42,6 +49,37 @@ suite('Telemetry', () => {
+        [
+            {
+                testName: 'Returns true when globalValue is set to false',
+                settings: { globalValue: false },
```
What about cases where `workspaceValue: false`, etc. are set?

karrtikr replied:
Currently we are not using `resource`, so only user settings are returned & settings will never contain `workspaceValue` as argument.

Even though it may not take much effort, I am not sure I need to enumerate possibilities in the test which do not exist functionality-wise as practice. @DonJayamanne @brettcannon

ericsnowcurrently commented on src/test/common/experiments.unit.test.ts:
```diff
@@ -0,0 +1,371 @@
+import { sleep } from '../common';
+
+// tslint:disable-next-line: max-func-body-length
+suite('A/B experiments', () => {
```
This would all be a bit easier to follow if you had a separate suite for each of the methods you are testing.

karrtikr replied:
Noted

ericsnowcurrently commented on src/test/common/experiments.unit.test.ts:
```diff
@@ -0,0 +1,371 @@
+            .setup(n => n.value)
+            .returns(() => undefined)
+            .verifiable(TypeMoq.Times.once());
+        when(fs.fileExists(anything())).thenResolve(false);
```
Please put each chained call on its own line.  It *really* helps with readability. :)

I'm only going to put this one comment, but please do it in all the applicable places.  Thanks!

DonJayamanne replied:
I disagree, this is something for the formatter to do.

ericsnowcurrently commented on src/test/common/experiments.unit.test.ts:
```diff
@@ -0,0 +1,371 @@
+        // First activation
+        await expManager.activate();
+
+        verify(fs.fileExists(anything())).once();
```
Checks for the first `activate()` call don't belong in a test for a second call.  This block of checks should be happen at the end of the test, *after* the second call.

ericsnowcurrently commented on src/test/common/experiments.unit.test.ts:
```diff
@@ -0,0 +1,371 @@
+        downloadedExperimentsStorage
+            .setup(n => n.value)
+            .returns(() => [{ name: 'experiment1', salt: 'salt', min: 90, max: 100 }])
+            .verifiable(TypeMoq.Times.atLeastOnce());
```
Please only use `atLeastOnce()` if the exact number may change between test runs.  In this case you should know the exact number.

This applies to all the places you are using `atLeastOnce()`.

ericsnowcurrently commented on src/test/common/experiments.unit.test.ts:
```diff
@@ -0,0 +1,371 @@
+        crypto = mock(CryptoUtils);
+        appEnvironment = mock(ApplicationEnvironment);
+        persistentStateFactory = mock(PersistentStateFactory);
+        isStorageValid = TypeMoq.Mock.ofType<IPersistentState<boolean>>();
```
Please use `TypeMoq.Mock.ofType<...>(undefined, TypeMoq.MockBehavior.Strict)`.  Doing so helps catch unexpected calls.

Also, be sure to call `verifyAll()` on all mocks in your tests.  Otherwise problems can slip through.

ericsnowcurrently commented on src/test/common/experiments.unit.test.ts:
```diff
@@ -0,0 +1,371 @@
+        experimentStorage
+            .setup(n => n.updateValue([{ name: 'experiment1', salt: 'salt', min: 90, max: 100 }]))
+            .returns(() => Promise.resolve(undefined)).verifiable(TypeMoq.Times.once());
+
```
Please check that `experimentStorage.value` is never called.  Using "String" mode for mocks, as suggested above, will take care of that for you.

karrtikr replied:
Refactored code doesn't have this issue anymore

karrtikr commented on src/client/activation/activationManager.ts:
```diff
@@ -26,7 +26,7 @@ export class ExtensionActivationManager implements IExtensionActivationManager {
         @inject(IApplicationDiagnostics) private readonly appDiagnostics: IApplicationDiagnostics,
         @inject(IWorkspaceService) private readonly workspaceService: IWorkspaceService
-    ) {}
+    ) { }
```
These are just format on save changes. I don't think there's anything wrong with it.

karrtikr commented on src/client/common/experiments.ts:
```diff
@@ -0,0 +1,154 @@
+        @inject(IOutputChannel) @named(STANDARD_OUTPUT_CHANNEL) private readonly output: IOutputChannel,
+        @inject(IFileSystem) private readonly fs: IFileSystem
+    ) {
+        this.isStorageValid = this.persistentStateFactory.createGlobalPersistentState<boolean>(isStorageValidKey, false, EXPIRY_DURATION_MS);
```
Experiments can be downloaded but expired, so `haveDownloadedExperiements` does not seem appropriate.
I think `haveValidDownloadedExperiements` or better `isDownloadedStorageValid` fits well with the expiry aspect of it.

karrtikr commented on src/client/common/types.ts:
```diff
@@ -407,7 +408,28 @@ export interface IAsyncDisposable {
+*/
+export type ABExperiments = { name: string; salt: string; min: number; max: number }[];
+export const IExperimentsManager = Symbol('IExperimentsManager');
+export interface IExperimentsManager {
```
Done.. although adding docstrings for both interfaces and classes feels like an overkill. @DonJayamanne @brettcannon @ericsnowcurrently 
If we have only one class implementing an interface, do we need to add it?

ericsnowcurrently commented on environments.json:
```diff
@@ -0,0 +1,6 @@
+[{
```
Why store this in a file and not just store it as a constant in `src/client/common/experiments.ts`?

karrtikr replied:
This file serves two purposes:
- As i have mentioned in the comments earlier, this is done to ensure that experiments are used in the first session itself, as about 40% of the users never come back for the second session. A constant won't work as it can only be used for the second session.

- The extension will be downloading this file from `https://raw.githubusercontent.com/microsoft/vscode-python/master/environments.json`

ericsnowcurrently replied:
I mean that we would hard-code the JSON data as a constant in experiments.ts.  This would be equivalent to having including the .json file with each release.  It would have the advantage that nobody is going to delete or modify the constant.

karrtikr replied:
We require a link to download the file from, so we have to commit the file anyways. Why not simply use it when the file is already there?

karrtikr replied:
So if a user decides to delete the experiment file? I don't think we should worry about such cases.
Besides we require a link to download the file from, so we have to commit the file anyways. Why not simply use it when the file is already there?

karrtikr replied:
Besides, the file is deleted just after the first activation.

DonJayamanne commented on src/client/common/net/httpClient.ts:
```diff
@@ -24,19 +24,28 @@ export class HttpClient implements IHttpClient {
     }
 
-    public async getJSON<T>(uri: string, strict: boolean = true): Promise<T> {
+    public request() {
```
You can remove this method.

DonJayamanne commented on src/client/common/net/httpClient.ts:
```diff
@@ -24,19 +24,28 @@ export class HttpClient implements IHttpClient {
+    }
+
+    public async getJSON<T>(uri: string, strict: boolean = true): Promise<T> {
+        const requestz = this.request();
```
You can change this to:
```typescript
const request = require('request') as typeof requestTypes;
```

DonJayamanne commented on src/test/common/net/httpClient.unit.test.ts:
```diff
@@ -1,35 +1,51 @@
+        workSpaceService.verifyAll();
+    });
+    test('Test getJSON()', async () => {
+        const request = sandbox.stub(HttpClient.prototype, 'request');
```
You can use `rewiremock` instead of creating a method `request`.
We use this today, its much easier. I believe its cleaner as well.

karrtikr replied:
Alright, will check it out.

DonJayamanne commented on .vscode/launch.json:
```diff
@@ -131,6 +128,27 @@
             "preLaunchTask": "Compile"
         },
+        {
+            "name": "Tests (VS Code, with code coverage, *.test.ts)",
```
Why are we adding this?
Please rebase against master

karrtikr commented:
@ericsnowcurrently I have addressed most of your comments. AFAICS,
* Test for empty experiments array, 1 experiment, multiple experiments
--- As length of the array does not matter functionally as of now, so I skip these
* CryptoUtils.createHash()
--- [ ] different values for the data arg
--- [ ] different values for the encoding arg
These values are directly associated with external library. I don't have the input and output to test this, plus I think I should only be testing the extension code and not an external library.
